File: explicit-java/src/main/kotlin/explicit/token/Slot.kt
package explicit.token

import explicit.api.IToken
import explicit.api.Navigator
import explicit.api.extension.isNumeric
import explicit.evaluator.ExplicitRuleEvaluation
import explicit.parser.ExplicitRules
import java.util.*

/**
 * Token implementation which represents a ner slot.
 */
data class Slot(val slot: String): IToken {
    override fun toString() = "{$slot}"
    override fun asText() = slot

    override fun evaluate(rulez: ExplicitRules, navigator: Navigator<String>): ExplicitRuleEvaluation {
        val expr1 = toString().equals(navigator.getCurr(), true)
        val expr2 = asText() == "number" && navigator.getCurr().isNumeric()
        val expr3 = evaluatePattern(rulez, navigator.getCurr())
        val expr4 = evaluateMapping(rulez, navigator.getCurr())

        val evaluation = expr1 || expr2 || expr3 || expr4
        return ExplicitRuleEvaluation(evaluation, HashMap(), listOf(navigator.getIndex() - 1))
    }

    private fun evaluatePattern(rulez: ExplicitRules, atomic: String): Boolean {
        if (rulez.patterns.containsKey(asText())) {
            val matcher = rulez.patterns[asText()]!!.matcher(atomic)
            if (matcher.find()) {
                return atomic.length == matcher.end()
            }
        }
        return false
    }

    private fun evaluateMapping(rulez: ExplicitRules, atomic: String): Boolean {
        if (rulez.mappings.containsKey(atomic)) {
            when (asText()) {
                "number" -> return rulez.mappings[atomic]?.isNumeric() ?: false
            }
        }
        return false
    }
}

File: explicit-python/explicit_nlu/token/Slot.py
import logging
import re
from dataclasses import dataclass

from explicit_nlu.api import IToken, Navigator
from explicit_nlu.api.support import Strings
from explicit_nlu.evaluator import ExplicitEvaluation
from explicit_nlu.parser import ExplicitRules


#
# Token implementation which represents a ner slot.
#
@dataclass
class Slot(IToken):
    slot: str

    def __str__(self):
        return "{" + self.slot + "}"

    def __repr__(self):
        return self.__str__()

    def as_text(self):
        return self.slot

    def evaluate(self, rulez: ExplicitRules, navigator: Navigator) -> ExplicitEvaluation:
        logging.debug("evaluate slot")

        expr1 = self.__str__() == navigator.get_curr()
        expr2 = self.as_text() == "number" and Strings.is_numeric(navigator.get_curr())
        expr3 = self.evaluate_pattern(rulez, navigator.get_curr())
        expr4 = self.evaluate_mapping(rulez, navigator.get_curr())

        evaluation = expr1 or expr2 or expr3 or expr4
        return ExplicitEvaluation(evaluation, dict(), [navigator.get_absolute_index() - 1])

    def evaluate_pattern(self, rulez: ExplicitRules, atomic: str) -> bool:
        if self.as_text() in rulez.patterns:
            match = re.match(rulez.patterns[self.as_text()] + "$", atomic)
            return match is not None
        return False

    def evaluate_mapping(self, rulez: ExplicitRules, atomic: str) -> bool:
        if atomic in rulez.mappings:
            if self.as_text() == "number":
                return rulez.mappings[atomic].isdigit()
        return False
